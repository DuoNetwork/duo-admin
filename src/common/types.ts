import { ChildProcess } from 'child_process';
export * from '../../../duo-contract-wrapper/src/types';

export interface IBaseMarketData {
	source: string;
	base: string;
	quote: string;
	timestamp: number;
}

export interface ITrade extends IBaseMarketData {
	id: string;
	price: number;
	amount: number;
}

export interface IPriceFix extends IBaseMarketData {
	price: number;
	volume: number;
}

export interface IPrice extends IBaseMarketData {
	period: number;
	open: number;
	high: number;
	low: number;
	close: number;
	volume: number;
}

export interface IBaseEvent {
	contractAddress: string;
	timestamp: number;
	transactionHash: string;
	blockNumber: number;
}

export interface IAcceptedPrice extends IBaseEvent {
	price: number;
	navA: number;
	navB: number;
}

export interface ITotalSupply extends IBaseEvent {
	tokenA: number;
	tokenB: number;
}

export interface IConversion extends IBaseEvent {
	type: string;
	eth: number;
	tokenA: number;
	tokenB: number;
	fee: number;
	pending?: boolean;
	reverted?: boolean;
}

export interface IStatus {
	process: string;
	timestamp: number;
}

export interface IPriceStatus extends IStatus {
	price: number;
	volume: number;
}

export interface INodeStatus extends IStatus {
	block: number;
}

export interface IAccount {
	address: string;
	privateKey: string;
}

export interface IKey {
	publicKey: string;
	privateKey: string;
}

export interface ISqlAuth {
	host: string;
	user: string;
	password: string;
}

export interface IOption {
	forceREST: boolean;
	pair: string;
	sources: string[];
	exSources: string[];
	assets: string[];
	live: boolean;
	dbLive: boolean;
	server: boolean;
	dynamo: boolean;
	aws: boolean;
	gcp: boolean;
	azure: boolean;
	force: boolean;
	gasPrice: number;
	gasLimit: number;
	source: string;
	event: string;
	provider: string;
	period: number;
	base: string;
	quote: string;
	contractType: string;
	tenor: string;
}

export interface ISourceSettings {
	priceInversed: boolean;
	quoteInversed: boolean;
	separator: string;
	isLowercase: boolean;
	tradesInterval: number;
	supportWS: boolean;
	filterByTimestamp: boolean;
	wsLink: string;
}

export interface ISourceAsset {
	base: {
		[code: string]: boolean;
	};
	quote: {
		[code: string]: boolean;
	};
	mapping: string;
	name: string;
}
export interface ISource {
	assets: {
		[code: string]: ISourceAsset;
	};
	settings: ISourceSettings;
}

export interface IBitfinexRawTradeWS {
	timestamp: number;
	seq: string;
	id: number;
	price: string;
	amount: string;
}

export interface IBitstampRawTradeRest {
	date: number;
	tid: number;
	price: number;
	type: string;
	amount: number;
}

export interface IBitstampRawTradeWs {
	microtimestamp: string;
	amount: number;
	buy_order_id: number;
	sell_order_id: number;
	amount_str: string;
	price_str: string;
	timestamp: string;
	price: number;
	type: number;
	id: number;
}

export interface IBitfinexRawTradeRest {
	timestamp: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeRest {
	timestamp: number;
	timestampms: number;
	tid: number;
	price: string;
	amount: string;
	exchange: string;
	type: string;
}

export interface IGeminiTradeWs {
	type: string;
	eventid: number;
	socket_sequence: number;
	timestamp: number;
	timestampms: number;
	events: IGeminiTradeData[];
}

export interface IGeminiTradeData {
	type: string;
	tid: number;
	price: string;
	amount: string;
	makerSide: string;
}

export type IKrakenRawTrade = IKrakenRawTradeEntry & {
	last: string;
};

export interface IKrakenRawTradeEntry {
	[pair: string]: string[][];
}

export interface IGdaxRawTradeRest {
	error: string[];
	result: {
		[code: string]: Array<Array<string | number>>;
	};
}

export interface ISubProcess {
	source: string;
	instance: ChildProcess;
	lastFailTimestamp: number;
	failCount: number;
}
