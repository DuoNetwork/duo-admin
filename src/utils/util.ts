import moment from 'moment';
import * as CST from '../common/constants';
import { IOption } from '../common/types';
// import { Options } from 'aws-cli-js';

class Util {
	public logLevel: string = CST.LOG_INFO;

	private log(text: any, level?: string): void {
		if (level && CST.LOG_RANKING[this.logLevel] >= CST.LOG_RANKING[level])
			console.log(`${moment().format('HH:mm:ss.SSS')} [${level}]: ` + text);
	}

	public logInfo(text: any): void {
		this.log(text, CST.LOG_INFO);
	}

	public logDebug(text: any): void {
		this.log(text, CST.LOG_DEBUG);
	}

	public logError(text: any): void {
		this.log(text, CST.LOG_ERROR);
	}

	public composeQuery(paras: { [key: string]: any }): string {
		let query: string = '?';
		for (const key in paras) query += key + '=' + paras[key] + '&';
		query = query.slice(0, -1);
		return query;
	}

	public parseOptions(argv: string[]): IOption {
		const option = {
			forceREST: argv.includes('rest'),
			live: process.argv.includes('live'),
			dbLive: process.argv.includes('dbLive'),
			server: process.argv.includes('server'),
			dynamo: process.argv.includes('dynamo'),
			aws: process.argv.includes('aws'),
			gcp: process.argv.includes('gcp'),
			azure: process.argv.includes('azure'),
			force: process.argv.includes('force'),
			pair: '',
			contractType: 'Beethoven',
			tenor: 'Perpetual',
			assets: [''],
			sources: [''],
			exSources: [''],
			gasPrice: 5e9,
			gasLimit: 200000,
			source: '',
			event: '',
			provider: '',
			period: 1,
			base: 'USD',
			quote: 'ETH'
		};
		for (let i = 3; i < argv.length; i++) {
			const args = argv[i].split('=');
			switch (args[0]) {
				case 'exSources':
					option.exSources = args[1].split(',');
					break;
				case 'sources':
					option.sources = args[1].split(',');
					break;
				case 'pair':
					option.pair = args[1].replace('_', '|') || option.pair;
					break;
				case 'assets':
					option.assets = args[1].split(',');
					break;
				case 'gasPrice':
					option.gasPrice = Number(args[1]) || option.gasPrice;
					break;
				case 'gasLimit':
					option.gasLimit = Number(args[1]) || option.gasLimit;
					break;
				case 'source':
					option.source = args[1] || option.source;
					break;
				case 'event':
					option.event = args[1] || option.event;
					break;
				case 'provider':
					option.provider = args[1] || option.provider;
					break;
				case 'period':
					option.period = Number(args[1]);
					break;
				case 'base':
					option.base = args[1] || option.base;
					break;
				case 'quote':
					option.quote = args[1] || option.quote;
					break;
				case 'contractType':
					option.contractType = args[1] || option.contractType;
					break;
				case 'tenor':
					option.tenor = args[1] || option.tenor;
					break;
				default:
					break;
			}
		}

		return option;
	}

	public getPeriodStartTimestamp(timestamp: number, period: number = 1) {
		// const now = this.getUTCNowTimestamp();
		return Math.floor(timestamp / 60000 / period - 1) * 60000 * period;
	}

	public timestampToString(ts: number) {
		//just for debugging purpose
		return moment.utc(ts).format('YYYY-MM-DDTHH:mm:ss');
	}

	public getStatusProcess(tool: string, option: IOption) {
		let type = '';
		const platform = option.azure ? '_AZURE' : option.gcp ? '_GCP' : '_AWS';
		console.log('dynamo', option.dynamo);
		const privacy = option.dynamo ? '_PUBLIC' : '_PRIVATE';
		let source = '';

		switch (tool) {
			case CST.TRADES:
				type = 'TRADE';
				source = option.source ? '_' + option.source.toUpperCase() : '';
				break;
			case CST.TRIGGER:
			case CST.FETCH_EVENTS:
				type = 'EVENT';
				source =
					'_' +
					([CST.EVENT_START_PRE_RESET, CST.EVENT_START_RESET].includes(option.event)
						? option.event.toUpperCase()
						: 'OTHERS');
				break;
			case CST.COMMIT:
				type = 'FEED';
				break;
			case CST.CLEAN_DB:
				type = 'CLEANDB';
				break;
			case CST.DB_PRICES:
				type = 'PRICE';
				if (option.period === 1) source = '_MINUTELY';
				else if (option.period === 60) source = '_HOURLY';
				else source = '';
				break;
			case 'minutely':
				type = 'MINUTELY';
				break;
			case CST.FETCH_PRICE:
				type = 'FETCH';
				break;
			default:
				return '';
		}

		return type + platform + privacy + source;
	}

	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public isNumber(input: any): boolean {
		const num = Number(input);
		return isFinite(num) && !isNaN(num);
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}

	public async sleep(ms: number) {
		return new Promise(resolve => {
			setTimeout(resolve, ms);
		});
	}
}

const util = new Util();
export default util;
