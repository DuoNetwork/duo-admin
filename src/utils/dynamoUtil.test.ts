import * as CST from '../common/constants';
import { IEvent } from '../common/types';
import conversion from '../samples/dynamo/conversion.json';
import prices from '../samples/dynamo/prices.json';
import status from '../samples/dynamo/status.json';
import totalSupply from '../samples/dynamo/totalSupply.json';
import uiCreate from '../samples/dynamo/uiCreate.json';
import uiRedeem from '../samples/dynamo/uiRedeem.json';
import dynamoUtil from './dynamoUtil';
import util from './util';

const trade = {
	quote: 'quote',
	base: 'base',
	source: 'src',
	id: 'id',
	price: 123,
	amount: 456,
	timestamp: 1234567890
};

const events: IEvent[] = [
	{
		type: 'type',
		id: 'id',
		blockHash: 'blockHash',
		blockNumber: 123,
		transactionHash: 'txHash',
		logStatus: 'logStatus',
		parameters: {
			test: 'test'
		},
		timestamp: 1234567890
	},
	{
		type: 'CommitPrice',
		id: 'id',
		blockHash: 'blockHash',
		blockNumber: 123,
		transactionHash: 'txHash',
		logStatus: 'logStatus',
		parameters: {
			sender: '0x00D8d0660b243452fC2f996A892D3083A903576F'
		},
		timestamp: 1234567890
	}
];

test('connection initalization', () =>
	dynamoUtil.insertData({} as any).catch(error => expect(error).toMatchSnapshot()));

test('convertTradeToDynamo', () =>
	expect(dynamoUtil.convertTradeToDynamo(trade, 123)).toMatchSnapshot());

test('convertEventToDynamo', () =>
	events.forEach(event =>
		expect(dynamoUtil.convertEventToDynamo(event, 9876543210)).toMatchSnapshot()
	));

test('insertTradeData', async () => {
	dynamoUtil.insertData = jest.fn(() => Promise.resolve());
	util.getNowTimestamp = jest.fn(() => 0);
	await dynamoUtil.insertTradeData(trade, true);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls.length).toBe(2);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
});

test('insertEventsData', async () => {
	dynamoUtil.insertData = jest.fn(() => Promise.resolve({}));
	await dynamoUtil.insertEventsData(events);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls.length).toBe(2);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
});

test('insertHeartbeat', async () => {
	dynamoUtil.insertData = jest.fn(() => Promise.resolve());
	await dynamoUtil.insertHeartbeat();
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls.length).toBe(1);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

test('insertStatusData', async () => {
	dynamoUtil.insertData = jest.fn(() => Promise.resolve());
	await dynamoUtil.insertStatusData({ test: 'test' });
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls.length).toBe(1);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

test('scanStatus', async () => {
	dynamoUtil.scanData = jest.fn(() => Promise.resolve({}));
	await dynamoUtil.scanStatus();
	expect((dynamoUtil.scanData as jest.Mock<Promise<void>>).mock.calls.length).toBe(1);
	expect((dynamoUtil.scanData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});

const convertedStatus = dynamoUtil.parseStatus(status);
test('parseStatus', () => expect(convertedStatus).toMatchSnapshot());

test('queryAcceptPriceEvent', async () => {
	dynamoUtil.queryData = jest.fn(() => Promise.resolve({}));
	await dynamoUtil.queryAcceptPriceEvent(['date1', 'date2']);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls.length).toBe(2);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
});

test('parseAcceptedPrices', () => expect(dynamoUtil.parseAcceptedPrice(prices)).toMatchSnapshot());

test('queryConversionEvent', async () => {
	dynamoUtil.queryData = jest.fn(() => Promise.resolve({}));
	await dynamoUtil.queryConversionEvent(CST.DUMMY_ADDR, ['date1', 'date2']);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls.length).toBe(4);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[2][0]).toMatchSnapshot();
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[3][0]).toMatchSnapshot();
});

test('parseConversion', () => expect(dynamoUtil.parseConversion(conversion)).toMatchSnapshot());

test('queryTotalSupplyEvent', async () => {
	dynamoUtil.queryData = jest.fn(() => Promise.resolve({}));
	await dynamoUtil.queryTotalSupplyEvent(['date1', 'date2']);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls.length).toBe(2);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
});

test('parseTotalSupply', () => expect(dynamoUtil.parseTotalSupply(totalSupply)).toMatchSnapshot());

test('queryUIConversionEvent', async () => {
	dynamoUtil.queryData = jest.fn(() => Promise.resolve({}));
	await dynamoUtil.queryUIConversionEvent(CST.DUMMY_ADDR);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls.length).toBe(2);
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.queryData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
});

test('parseUIConversion', () => {
	expect(dynamoUtil.parseUIConversion(uiCreate)).toMatchSnapshot();
	expect(dynamoUtil.parseUIConversion(uiRedeem)).toMatchSnapshot();
});

test('insertUIConversion', async () => {
	dynamoUtil.insertData = jest.fn(() => Promise.resolve());
	util.getNowTimestamp = jest.fn(() => 1234567890);
	await dynamoUtil.insertUIConversion(CST.DUMMY_ADDR, '0x123', true, 123, 456, 456, 0.123, 0);
	await dynamoUtil.insertUIConversion(CST.DUMMY_ADDR, '0x123', false, 123, 456, 456, 0, 0.123);
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
	expect((dynamoUtil.insertData as jest.Mock<Promise<void>>).mock.calls[1][0]).toMatchSnapshot();
});

test('deleteUIConversionEvent', async () => {
	dynamoUtil.deleteData = jest.fn(() => Promise.resolve());
	await dynamoUtil.deleteUIConversionEvent(CST.DUMMY_ADDR, {
		type: 'type',
		transactionHash: 'txHash',
		eth: 0,
		tokenA: 0,
		tokenB: 0,
		timestamp: 0,
		blockNumber: 0,
		ethFee: 0,
		duoFee: 0
	});
	expect((dynamoUtil.deleteData as jest.Mock<Promise<void>>).mock.calls[0][0]).toMatchSnapshot();
});
