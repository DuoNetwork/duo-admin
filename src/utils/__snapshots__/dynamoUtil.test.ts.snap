// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connection initalization 1`] = `"dynamo db connection is not initialized"`;

exports[`convertEventToDynamo 1`] = `
Object {
  "blockHash": Object {
    "S": "blockHash",
  },
  "blockNumber": Object {
    "N": "123",
  },
  "eventKey": Object {
    "S": "type|1970-01-15",
  },
  "logStatus": Object {
    "S": "logStatus",
  },
  "systime": Object {
    "N": "9876543210",
  },
  "test": Object {
    "S": "test",
  },
  "timestampId": Object {
    "S": "1234567890|id",
  },
  "transactionHash": Object {
    "S": "txHash",
  },
}
`;

exports[`convertEventToDynamo 2`] = `
Object {
  "blockHash": Object {
    "S": "blockHash",
  },
  "blockNumber": Object {
    "N": "123",
  },
  "eventKey": Object {
    "S": "CommitPrice|1970-01-15",
  },
  "logStatus": Object {
    "S": "logStatus",
  },
  "sender": Object {
    "S": "0x00D8d0660b243452fC2f996A892D3083A903576F",
  },
  "systime": Object {
    "N": "9876543210",
  },
  "timestampId": Object {
    "S": "1234567890|id",
  },
  "transactionHash": Object {
    "S": "txHash",
  },
}
`;

exports[`convertTradeToDynamo 1`] = `
Object {
  "amount": Object {
    "N": "456",
  },
  "price": Object {
    "N": "123",
  },
  "quoteBaseId": Object {
    "S": "quote|base|id",
  },
  "systime": Object {
    "N": "123",
  },
  "timestamp": Object {
    "N": "1234567890",
  },
}
`;

exports[`deleteUIConversionEvent 1`] = `
Object {
  "Key": Object {
    "eventKey": Object {
      "S": "type|0x0",
    },
    "transactionHash": Object {
      "S": "txHash",
    },
  },
  "TableName": "duo.uiEvents.dev",
}
`;

exports[`insertEventsData 1`] = `
Object {
  "Item": Object {
    "blockHash": Object {
      "S": "blockHash",
    },
    "blockNumber": Object {
      "N": "123",
    },
    "eventKey": Object {
      "S": "type|1970-01-15",
    },
    "logStatus": Object {
      "S": "logStatus",
    },
    "systime": Object {
      "N": "0",
    },
    "test": Object {
      "S": "test",
    },
    "timestampId": Object {
      "S": "1234567890|id",
    },
    "transactionHash": Object {
      "S": "txHash",
    },
  },
  "TableName": "duo.events.dev",
}
`;

exports[`insertEventsData 2`] = `
Object {
  "Item": Object {
    "blockHash": Object {
      "S": "blockHash",
    },
    "blockNumber": Object {
      "N": "123",
    },
    "eventKey": Object {
      "S": "CommitPrice|1970-01-15",
    },
    "logStatus": Object {
      "S": "logStatus",
    },
    "sender": Object {
      "S": "0x00D8d0660b243452fC2f996A892D3083A903576F",
    },
    "systime": Object {
      "N": "0",
    },
    "timestampId": Object {
      "S": "1234567890|id",
    },
    "transactionHash": Object {
      "S": "txHash",
    },
  },
  "TableName": "duo.events.dev",
}
`;

exports[`insertHeartbeat 1`] = `
Object {
  "Item": Object {
    "process": Object {
      "S": "UNKNOWN",
    },
    "timestamp": Object {
      "N": "0",
    },
  },
  "TableName": "duo.status.dev",
}
`;

exports[`insertStatusData 1`] = `
Object {
  "Item": Object {
    "process": Object {
      "S": "UNKNOWN",
    },
    "test": "test",
  },
  "TableName": "duo.status.dev",
}
`;

exports[`insertTradeData 1`] = `
Object {
  "Item": Object {
    "amount": Object {
      "N": "456",
    },
    "price": Object {
      "N": "123",
    },
    "quoteBaseId": Object {
      "S": "quote|base|id",
    },
    "sourceDateHourMinute": Object {
      "S": "src|1970-01-15-06-56",
    },
    "systime": Object {
      "N": "0",
    },
    "timestamp": Object {
      "N": "1234567890",
    },
  },
  "TableName": "duo.trades.dev",
}
`;

exports[`insertTradeData 2`] = `
Object {
  "Item": Object {
    "amount": Object {
      "N": "456",
    },
    "price": Object {
      "N": "123",
    },
    "process": Object {
      "S": "UNKNOWN",
    },
    "quoteBaseId": Object {
      "S": "quote|base|id",
    },
    "systime": Object {
      "N": "0",
    },
    "timestamp": Object {
      "N": "1234567890",
    },
  },
  "TableName": "duo.status.dev",
}
`;

exports[`insertUIConversion 1`] = `
Object {
  "Item": Object {
    "duoFee": Object {
      "N": "0",
    },
    "eth": Object {
      "N": "123",
    },
    "ethFee": Object {
      "N": "0.123",
    },
    "eventKey": Object {
      "S": "Create|0x0",
    },
    "systime": Object {
      "N": "1234567890",
    },
    "tokenA": Object {
      "N": "456",
    },
    "tokenB": Object {
      "N": "456",
    },
    "transactionHash": Object {
      "S": "0x123",
    },
  },
  "TableName": "duo.uiEvents.dev",
}
`;

exports[`insertUIConversion 2`] = `
Object {
  "Item": Object {
    "duoFee": Object {
      "N": "0.123",
    },
    "eth": Object {
      "N": "123",
    },
    "ethFee": Object {
      "N": "0",
    },
    "eventKey": Object {
      "S": "Redeem|0x0",
    },
    "systime": Object {
      "N": "1234567890",
    },
    "tokenA": Object {
      "N": "456",
    },
    "tokenB": Object {
      "N": "456",
    },
    "transactionHash": Object {
      "S": "0x123",
    },
  },
  "TableName": "duo.uiEvents.dev",
}
`;

exports[`parseAcceptedPrices 1`] = `
Array [
  Object {
    "blockNumber": 7722428,
    "navA": 0,
    "navB": 0,
    "price": 0,
    "timestamp": 1529625600000,
    "transactionHash": "0x180245f592adfd86dfb49a48c6c9b59d6be7a8d9c897ebd5492f0823c5196008",
  },
  Object {
    "blockNumber": 7722781,
    "navA": 0,
    "navB": 0,
    "price": 0,
    "timestamp": 1529629200000,
    "transactionHash": "0x4da754a05a29dd84a26a81db5a6d437be018971c652a94d0c706128bad63dc87",
  },
  Object {
    "blockNumber": 7723130,
    "navA": 0,
    "navB": 0,
    "price": 0,
    "timestamp": 1529632800000,
    "transactionHash": "0x298dc4d2771805df61894777a6a2eee702d222ab51c1c5239d890d66412b64e0",
  },
  Object {
    "blockNumber": 7723483,
    "navA": 0,
    "navB": 0,
    "price": 0,
    "timestamp": 1529636400000,
    "transactionHash": "0xef0aab40c21251ddbed21d03f43f81a0e000b651a355dc2e0ac6bdd2934e449d",
  },
  Object {
    "blockNumber": 7723836,
    "navA": 0,
    "navB": 0,
    "price": 0,
    "timestamp": 1529640000000,
    "transactionHash": "0x060f2496b45b7d11908549bb4e3e1a6cc789692caf24c9f632daf5e23ce3a481",
  },
]
`;

exports[`parseConversion 1`] = `
Array [
  Object {
    "blockNumber": 7715602,
    "duoFee": 0,
    "eth": 0,
    "ethFee": 0,
    "timestamp": 1529559912000,
    "tokenA": 0,
    "tokenB": 0,
    "transactionHash": "0x7283db82115b77c3fa055ec1553acf7c16d3ba1e2db5b2cea1974ba32bad29dc",
    "type": "Create",
  },
]
`;

exports[`parseStatus 1`] = `
Array [
  Object {
    "block": 7535634,
    "process": "CHAIN_AWS_PUBLIC",
    "timestamp": 1527846160749,
  },
  Object {
    "process": "EVENT_AWS_PRIVATE_STARTPRERESET",
    "timestamp": 1527751553589,
  },
  Object {
    "process": "EVENT_AWS_PRIVATE_STARTRESET",
    "timestamp": 1527751833800,
  },
  Object {
    "process": "EVENT_AZURE_PRIVATE_STARTPRERESET",
    "timestamp": 1527752331975,
  },
  Object {
    "process": "EVENT_AZURE_PRIVATE_STARTRESET",
    "timestamp": 1527751947443,
  },
  Object {
    "process": "EVENT_GCP_PRIVATE_STARTPRERESET",
    "timestamp": 1527752140470,
  },
  Object {
    "process": "EVENT_GCP_PRIVATE_STARTRESET",
    "timestamp": 1527752036379,
  },
  Object {
    "process": "FEED_AWS_PRIVATE",
    "timestamp": 1527851700049,
  },
  Object {
    "process": "FEED_AZURE_PRIVATE",
    "timestamp": 1527758100003,
  },
  Object {
    "process": "FEED_GCP_PRIVATE",
    "timestamp": 1527757500018,
  },
  Object {
    "process": "HOURLY_AWS_PUBLIC",
    "timestamp": 1528081753755,
  },
  Object {
    "process": "MINUTELY_AWS_PUBLIC",
    "timestamp": 1528081764433,
  },
  Object {
    "price": 573.84,
    "process": "PRICE_AWS_PRIVATE_BITFINEX",
    "timestamp": 1527831872000,
    "volume": 2,
  },
  Object {
    "price": 568.74,
    "process": "PRICE_AWS_PRIVATE_GDAX",
    "timestamp": 1527756997255,
    "volume": 0.112,
  },
  Object {
    "price": 568.9,
    "process": "PRICE_AWS_PRIVATE_GEMINI",
    "timestamp": 1527757021801,
    "volume": 0.034274,
  },
  Object {
    "price": 567.99,
    "process": "PRICE_AWS_PRIVATE_KRAKEN",
    "timestamp": 1527756882730,
    "volume": 1.3,
  },
  Object {
    "price": 617.38,
    "process": "PRICE_AWS_PUBLIC_BITFINEX",
    "timestamp": 1528081780000,
    "volume": 1.608,
  },
  Object {
    "price": 616.14,
    "process": "PRICE_AWS_PUBLIC_GDAX",
    "timestamp": 1528081164613,
    "volume": 6.52551355,
  },
  Object {
    "price": 617.67,
    "process": "PRICE_AWS_PUBLIC_GEMINI",
    "timestamp": 1528081767414,
    "volume": 0.021854,
  },
  Object {
    "price": 617.85,
    "process": "PRICE_AWS_PUBLIC_KRAKEN",
    "timestamp": 1528081700549,
    "volume": 1.13,
  },
  Object {
    "price": 569.89,
    "process": "PRICE_AZURE_PRIVATE_BITFINEX",
    "timestamp": 1527757209000,
    "volume": 0.02548273,
  },
  Object {
    "price": 569.09,
    "process": "PRICE_AZURE_PRIVATE_GDAX",
    "timestamp": 1527757217889,
    "volume": 0.69683,
  },
  Object {
    "price": 569.01,
    "process": "PRICE_AZURE_PRIVATE_GEMINI",
    "timestamp": 1527757197779,
    "volume": 0.714947,
  },
  Object {
    "price": 569.3,
    "process": "PRICE_AZURE_PRIVATE_KRAKEN",
    "timestamp": 1527757210567,
    "volume": 0.29453729,
  },
  Object {
    "price": 575.97,
    "process": "PRICE_GCP_PRIVATE_BITFINEX",
    "timestamp": 1527766704000,
    "volume": 5.16121098,
  },
  Object {
    "price": 575,
    "process": "PRICE_GCP_PRIVATE_GDAX",
    "timestamp": 1527766704262,
    "volume": 0.7576,
  },
  Object {
    "price": 575.41,
    "process": "PRICE_GCP_PRIVATE_GEMINI",
    "timestamp": 1527766669797,
    "volume": 0.005219,
  },
  Object {
    "price": 575.04,
    "process": "PRICE_GCP_PRIVATE_KRAKEN",
    "timestamp": 1527766688094,
    "volume": 0.2,
  },
]
`;

exports[`parseTotalSupply 1`] = `
Array [
  Object {
    "blockNumber": 7650502,
    "timestamp": 1528862200000,
    "tokenA": 0,
    "tokenB": 0,
    "transactionHash": "0xd7c9bf35f925a8353032cfe1178439617997d6e27484465d393de3d38f8331ab",
  },
  Object {
    "blockNumber": 7651095,
    "timestamp": 1528867600000,
    "tokenA": 0,
    "tokenB": 0,
    "transactionHash": "0xe35137d392e20dd2025cd996de73a271ebcea23128a89f7f423b326ce20893e5",
  },
]
`;

exports[`parseUIConversion 1`] = `
Array [
  Object {
    "blockNumber": 0,
    "duoFee": 0,
    "eth": 0.21704519241528358,
    "ethFee": 0.0021704519241528357,
    "pending": true,
    "timestamp": 1529317121530,
    "tokenA": 36.9599025612444,
    "tokenB": 36.9599025612444,
    "transactionHash": "0x8e59dff2458a76b2ba0a9e11413e603cddc227d663a8bf7b45e0241f3896424d",
    "type": "Create",
  },
]
`;

exports[`parseUIConversion 2`] = `
Array [
  Object {
    "blockNumber": 0,
    "duoFee": 2.0051663085602844,
    "eth": 0.20051663085602844,
    "ethFee": 0,
    "pending": true,
    "timestamp": 1529317783500,
    "tokenA": 34.14530889109876,
    "tokenB": 34.14530889109876,
    "transactionHash": "0x94ea38c42188eaacfa75ad0a091866b46f4c75993d33c94c8cc22822433ca244",
    "type": "Redeem",
  },
]
`;

exports[`queryAcceptPriceEvent 1`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "AcceptPrice|date1",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryAcceptPriceEvent 2`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "AcceptPrice|date2",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryConversionEvent 1`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "Create|date1|0x0",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryConversionEvent 2`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "Redeem|date1|0x0",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryConversionEvent 3`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "Create|date2|0x0",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryConversionEvent 4`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "Redeem|date2|0x0",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryTotalSupplyEvent 1`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "TotalSupply|date1",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryTotalSupplyEvent 2`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "TotalSupply|date2",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.events.dev",
}
`;

exports[`queryUIConversionEvent 1`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "Create|0x0",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.uiEvents.dev",
}
`;

exports[`queryUIConversionEvent 2`] = `
Object {
  "ExpressionAttributeValues": Object {
    ":eventKey": Object {
      "S": "Redeem|0x0",
    },
  },
  "KeyConditionExpression": "eventKey = :eventKey",
  "TableName": "duo.uiEvents.dev",
}
`;

exports[`scanStatus 1`] = `
Object {
  "TableName": "duo.status.dev",
}
`;
